### PR TITLE
C51-254: Add basic extension structure

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview

This PR adds a basic structure for the CiviCase Extras.

## How it looks

screen shot 2018-10-12 at 9 58 08 am
Notice the "Civicase extras success" message on the console. This is a temporary message to make sure the extension works and can overwrite Civicase.

## Technical details

The extension was created using the [civix](https://docs.civicrm.org/dev/en/latest/extensions/civix/#generate-module) tool:

```
civix generate:module uk.co.compucorp.civicaseextras
```

The structure of the angular application remains the same as CiviCase because that's how `civix` bootstraps angular applications and makes the integration with CiviCase easier. The files will be placed in the `ang` folder.

To make changes to CiviCase, we need to add this extension as a requirement of the former. To accomplish this, the `angularModules` hook was refactored as follows:

```php
function _civicaseextras_civix_civicrm_angularModules(&$angularModules) {
  _civicaseextras_includeAngularModules($angularModules);
  _civicaseextras_addCiviCaseExtrasAsRequirementForCivicase($angularModules);
}

function _civicaseextras_includeAngularModules(&$angularModules) {
  // this is the old code for the `angularModules`. It includes any files that match the pattern `ang/*.ang.php`
}

// This is how this extension is added as a requirement of civicase:
function _civicaseextras_addCiviCaseExtrasAsRequirementForCivicase(&$angularModules) {
  if (isset($angularModules['civicase'])) {
    $angularModules['civicase']['requires'][] = 'civicaseextras';
  } else {
    CRM_Core_Session::setStatus(
      'The <strong>Civicase Extras</strong> extension requires <strong>CiviCase</strong> to be installed first.',
      'Warning',
      'no-popup'
    );
  }
}
```

### CODE REVIEW

File: LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,667 @@
+Package: uk.co.compucorp.civicaseextras
+Copyright (C) 2018, René Olivo <minet.ws@gmail.com>
```
Comment (deb1990) [inline LICENSE.txt]: Compucorp Copyright info should be used. Please use the same from civicase repo.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  "main": "index.js",
+  "scripts": {
+    "test": "npx gulp test"
+  },
+  "semistandard": {
+    "globals": [
+      "CRM",
+      "angular",
+      "inject",
+      "moment"
```
Comment (deb1990) [inline package.json]: We do not need `moment` yet.

File: civicaseextras.php [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+ *
+ * Generate a list of Angular modules.
+ *
+ * Note: This hook only runs in CiviCRM 4.5+. It may
+ * use features only available in v4.6+.
+ *
+ * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_angularModules
+ */
+function civicaseextras_civicrm_angularModules(&$angularModules) {
+  _civicaseextras_civix_civicrm_angularModules($angularModules);
```
Comment (deb1990) [inline civicaseextras.php:114]: `_civicaseextras_civix_civicrm_angularModules ` is not defined yet?

File: civicaseextras.php [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+/**
+ * Implements hook_civicrm_alterSettingsFolders().
+ *
+ * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_alterSettingsFolders
+ */
+function civicaseextras_civicrm_alterSettingsFolders(&$metaDataFolders = NULL) {
+  _civicaseextras_civix_civicrm_alterSettingsFolders($metaDataFolders);
```
Comment (deb1990) [inline civicaseextras.php:123]: Not defined yet?

**Review by deb1990 — CHANGES_REQUESTED** [review]

**Comment by deb1990** [comment]
@reneolivo PR should be made to `develop` branch

File: LICENSE.txt (hunk 1 above)
Comment (reneolivo) [reply LICENSE.txt]: This was done automatically, thanks for spotting it.

File: civicaseextras.php (hunk 3 above)
Comment (reneolivo) [reply civicaseextras.php:114]: yes, `_civicaseextras_civix_civicrm_angularModules` is defined in `civicaseextras.civix.php`:
https://github.com/compucorp/uk.co.compucorp.civicaseexrtas/pull/1/files#diff-3e362daca7fbcc9c78d00b63e88ef045R316

This function is first introduced by default, and then it was modified so `civicase` requires `civicaseextras`.

File: civicaseextras.php (hunk 4 above)
Comment (reneolivo) [reply civicaseextras.php:123]: Also defined in `civicaseextras.civix.php`:

https://github.com/compucorp/uk.co.compucorp.civicaseexrtas/pull/1/files#diff-3e362daca7fbcc9c78d00b63e88ef045R463

This was defined by the plugin by default, and no modification was made.

**Review by deb1990 — APPROVED** [review]

**Pull request merged**